puffin/src/scope_details.rs

```diff
@@ -0,0 +1,212 @@
+
+use parking_lot::RwLock;
+
+use crate::{clean_function_name, fetch_add_scope_id, short_file_name, ScopeId};
+
+#[derive(Default, Clone)]
+struct InnerDetails {
+    pub(crate) scope_id_to_details: std::collections::HashMap<ScopeId, ScopeDetailsOwned>,
+    pub(crate) string_to_scope_id: std::collections::HashMap<String, ScopeId>,
+}
+
+/// Provides read access to scope details.
+#[derive(Default, Clone)]
+pub struct ScopeDetails(
+    // Store a both-way map, memory wise this can be a bit redundant but allows for faster access of information.
+    Arc<RwLock<InnerDetails>>,
+);
+
+impl ScopeDetails {
+    /// Provides read to the given closure for some scope details.
+    pub fn read_by_id<F: FnMut(&ScopeDetailsOwned)>(&self, scope_id: &ScopeId, mut cb: F) {
+        if let Some(read) = self.0.read().scope_id_to_details.get(scope_id) {
+            cb(read);
+        }
+    }
+
+    /// Provides read to the given closure for some scope details.
+    pub fn read_by_name<F: FnMut(&ScopeId)>(&self, scope_name: &str, mut cb: F) {
+        if let Some(read) = self.0.read().string_to_scope_id.get(scope_name) {
+            cb(read);
+        }
+    }
+
+    /// Function is hidden as only puffin library will insert entries.
+    pub(crate) fn insert(&self, scope_id: ScopeId, scope_details: ScopeDetailsOwned) {
+        self.0
+            .write()
+            .string_to_scope_id
+            .insert(scope_details.scope_name.to_string(), scope_id);
+        self.0
+            .write()
+            .scope_id_to_details
+            .insert(scope_id, scope_details);
+    }
+
+    /// Manually insert scope details. After a scope is inserted it can be reported to puffin.
+    pub fn insert_custom_scopes(&self, scopes: &[CustomScopeDetails]) -> HashSet<ScopeId> {
+        let mut new_scopes = HashSet::new();
+        for scope_detail in scopes {
+            let new_scope_id = fetch_add_scope_id();
+
+            self.insert(new_scope_id, ScopeDetailsOwned::from(scope_detail));
+
+            new_scopes.insert(new_scope_id);
+        }
+
+        new_scopes
+    }
+
+    /// Fetches all registered scopes and their ids.
+    /// Useful for fetching scope id by a static scope name.
+    pub fn scopes_by_name<T>(
+        &self,
+        mut existing_scopes: impl FnMut(&std::collections::HashMap<String, ScopeId>) -> T,
+    ) -> T {
+        existing_scopes(&self.0.read().string_to_scope_id)
+    }
+
+    /// Fetches all registered scopes.
+    /// Useful for fetching scope details by a scope id.
+    pub fn scopes_by_id<T>(
+        &self,
+        mut existing_scopes: impl FnMut(&std::collections::HashMap<ScopeId, ScopeDetailsOwned>) -> T,
+    ) -> T {
+        existing_scopes(&self.0.read().scope_id_to_details)
+    }
+}
+
+/// Scope details that are registered by the macros.
+/// This is only used internally and should not be exposed.
+#[derive(Debug, Clone, Copy, PartialEq, Hash, PartialOrd, Ord, Eq)]
+pub struct ScopeDetailsRef {
```

Comment:
We should make it clear (by name and docstring) that this is the static information that doesn't change with each invocation.

Perhaps `StaticScopeInfo` would be a good name?